src/redisearch_rs/trie_rs/src/trie.rs

```
@@ -0,0 +1,891 @@
+
+        self.children = ChildRefs(children);
+    }
+
+    /// Remove a child from the node.
+    /// Returns the data associated with the key if it was present.
+    fn remove_child(&mut self, key: &[c_char]) -> Option<Data> {
+        // Find the index of child whose label starts with the first byte of the key,
+        // as well as the child itself.
+        // If the we find none, there's nothing to remove.
+        // Note that `key.first()?` will cause this function to return None if the key is empty.
+        let child_index = self.children.index_of(*key.first()?)?;
+        let child = &mut self.children.0[child_index].node;
+
+        // If the child's label is equal to the key, we remove the child.
+        if child.label == key {
+            let data = child.data.take();
+
+            let child_is_leaf = child.children.is_empty();
+
+            if child_is_leaf {
+                // If the child is a leaf, we remove the child node itself.
+                self.children.remove(child_index);
+            } else {
+                // If there's a single grandchild,
+                // we merge the grandchild into the child.
+                child.merge_child();
+            }
+
+            return data;
+        }
+
+        // If the child's label is prefixed by the key, we recurse into the child.
+        // If not, there's nothing to remove.
+        let suffix = key.strip_prefix(child.label.as_slice())?;
+        debug_assert!(!suffix.is_empty());
+
+        let data = child.remove_child(suffix);
+        child.merge_child();
+        data
+    }
+
+    /// If `self` has exactly one child, and `self` doesn't hold
+    /// any data, merge child into `self`, by moving the child's data and
+    /// childreninto `self`. Depending on the spare capacity of `self`s label
+    /// and that of the child, we either extend `self`s label with the child's label,
+    /// or vice versa.
+    fn merge_child(&mut self) {
+        if self.data.is_some() {
```

why is it okay to silently fail/not merge if the node has data?

---

I think it makes the code less clear.
I think the if should be removed from this function.
maybe a debug assert can be in its place.

perhaps create a `try_merge_child` function which checks the condition where the data is not empty and only then does it calls `merge_child`.

---

Clarified in 9a3b257efe2c029278a3e2ed69345fbdde937a1a